catlearn/relation_cache.py

```diff
@@ -492,3 +494,35 @@ def match(
                 result_graph[src][tar].pop(NegativeMatch(best_match), None)
 
         return result_graph
+
+    def sort_relations(
+            self,
+            src: Optional[NodeType] = None, tar: Optional[NodeType] = None,
+            labels: Optional[FrozenSet[ArrowType]] = None
+    ) -> OrderedDict[Tuple[NodeType, NodeType, FrozenSet[ArrowType]], float]:
+        """
+        Given a result graph from a match, sort available relations matching
+        the specific properties (in terms of score):
+        - if src is given, will only look at relations with this source
+        - if tar is given, will only look at relations with this target
+        - if labels is given, will only look at the score corresponding to
+        the sum for these labels. Otherwise will list all possible
+        individual label scores.
+        """
+        arrows = self.arrows(src, tar, include_non_causal=False)
+
+        if labels is None:
+            options = {
+                (arr[0], arr[-1], frozenset((label,))): float(kl_match(
```

Comment:
Are we sure we have unicity here? I mean can we have a single arrow a->b and a->c->b (causal) ? The second one would override the first one

Reply:
It's possible indeed. In particular if there is only one possible label in the dataset the a->b arrow will always have a better score than the a->b->c arrow, so it will have a better rank.
However if there are several labels the a->b->c arrow may be a better match than the a->b arrow, as the causality constrain does not say anything about which specific label an arrow holds.

Reply:
let's keep the simpler version as it is. if it does not work, possible shorter-is-better causality parameter can be added. Current dataset has average relation order from test to train around 4-5, cannot expect there will be too many direct relations to override.

Reply:
We really should fix it: please refer to offline proof on paper :D